test/unit/atomic.c

```diff
@@ -1,101 +1,243 @@
-		assert_##ta##_eq(atomic_sub_##p(&accum, tests[i].x),	\
-		    (t)((tc)tests[i].accum0 - (tc)tests[i].x),		\
-		    "i=%u, accum=%"FMT", x=%"FMT,			\
-		    i, tests[i].accum0, tests[i].x);			\
-		assert_##ta##_eq(atomic_read_##p(&accum), accum,	\
-		    "Erroneous sub, i=%u", i);				\
-									\
-		accum = tests[i].accum0;				\
-		err = atomic_cas_##p(&accum, tests[i].x, tests[i].s);	\
-		assert_b_eq(err, tests[i].accum0 != tests[i].x,		\
-		    "Erroneous cas success/failure result");		\
-		assert_##ta##_eq(accum, err ? tests[i].accum0 :		\
-		    tests[i].s, "Erroneous cas effect, i=%u", i);	\
-									\
-		accum = tests[i].accum0;				\
-		atomic_write_##p(&accum, tests[i].s);			\
-		assert_##ta##_eq(accum, tests[i].s,			\
-		    "Erroneous write, i=%u", i);			\
-	}								\
+/*
+ * t: the non-atomic type, like "uint32_t".
+ * ta: the short name for the type, like "u32".
+ * val[1,2,3]: Values of the given type.  The CAS tests use val2 for expected,
+ * and val3 for desired.
+ */
+
+#define DO_TESTS(t, ta, val1, val2, val3) do {                                 \
+	t val;                                                                 \
+	t raw_atomic;                                                          \
+	t expected;                                                            \
+	bool success;                                                          \
+	/* This (combined with the load below) also tests ATOMIC_LOAD. */      \
+	atomic_##ta##_t atom = ATOMIC_INIT(val1);                              \
+                                                                               \
+	/* ATOMIC_INIT and load. */                                            \
+	val = atomic_load_##ta(&atom, ATOMIC_RELAXED);                         \
+	assert_##ta##_eq(val1, val, "Load or init failed");                    \
+                                                                               \
+	/* Store. */                                                           \
+	atomic_store_##ta(&atom, val1, ATOMIC_RELAXED);                        \
+	atomic_store_##ta(&atom, val2, ATOMIC_RELAXED);                        \
+	val = atomic_load_##ta(&atom, ATOMIC_RELAXED);                         \
+	assert_##ta##_eq(val2, val, "Store failed");                           \
+                                                                               \
+	/* Exchange. */                                                        \
+	atomic_store_##ta(&atom, val1, ATOMIC_RELAXED);                        \
+	val = atomic_exchange_##ta(&atom, val2, ATOMIC_RELAXED);               \
+	assert_##ta##_eq(val1, val, "Exchange returned invalid value");        \
+	val = atomic_load_##ta(&atom, ATOMIC_RELAXED);                         \
+	assert_##ta##_eq(val2, val, "Exchange store invalid value");           \
+                                                                               \
+	/* Weak CAS.  Spurious failures are allowed, so we loop a few times. */\
```

Comment:
Not meaningfully. This only matters for LL/SC architectures in screwy situations (e.g. taking an interrupt in between the load and store components of the CAS). This is more a theoretical possibility than a real one.